### ISSUE TITLE
Accessing Identity output

### ISSUE THREAD

**Issue by jkonecki-techfabric** [pulumi/pulumi-azure-nextgen#115]
I'm trying to create a new VM ScaleSet with system-assigned identity and than create KeyVault access policy for this identity:

```
var scaleSet = new VirtualMachineScaleSet($"{tenant}-orleans-scaleset-{environment}", new VirtualMachineScaleSetArgs
{
	Identity = new VirtualMachineScaleSetIdentityArgs
	{
		Type = "SystemAssigned"
	}                
});

new AccessPolicy($"{tenant}-orleans-scaleset-{environment}", new AccessPolicyArgs
{
	KeyVaultId = keyVault.VaultId,
	ObjectId = scaleSet.Identity.Apply(x => x.PrincipalId),
	TenantId = scaleSet.Identity.Apply(x => x.TenantId),
	KeyPermissions = new[] { "get", "list" },
	SecretPermissions = new[] { "get", "list" }
});
```

When running `pulumi up` I'm getting the following error:
```
  azure:keyvault:AccessPolicy (vrm-orleans-scaleset-dev):
    error: azure:keyvault/accessPolicy:AccessPolicy resource 'vrm-orleans-scaleset-dev' has a problem: Missing required property 'tenantId'
    error: azure:keyvault/accessPolicy:AccessPolicy resource 'vrm-orleans-scaleset-dev' has a problem: Missing required property 'objectId'
```

Am I using `Output.Apply` correctly?

**Comment by mikhailshilkov** [comment]
Your `Apply` usage looks good.

1. Do I understand correctly that `AccessPolicy` is from `Pulumi.Azure`? (if so, I'll probably move the issue to that repo)
2. This sounds like [this issue](https://github.com/pulumi/pulumi-azure/issues/192) which was recently fixed. Which provider version are you using?

**Comment by jkonecki-techfabric** [comment]
Oh, Scaleset is from AzureNextGen and AccessPolicy is from Azure provider.

Would this course a problem? Can inputs/outputs link resources managed by different providers?
________________________________

**Comment by mikhailshilkov** [comment]
> Would this course a problem? Can inputs/outputs link resources managed by different providers?

This is not a problem, that's allowed.

**Comment by mikhailshilkov** [comment]
Although, on the second thought, it is related. The issue I mentioned above was fixed for output flow for the Azure provider. I guess something is still missing for Azure NextGen. You can work around the issue with this:

```
    ObjectId = scaleSet.Identity.Apply(x => x.PrincipalId ?? Guid.Empty.ToString()),
    TenantId = scaleSet.Identity.Apply(x => x.TenantId ?? Guid.Empty.ToString()),
```

**Comment by jkonecki-techfabric** [comment]
Workaround worked. Thank you.
________________________________

**Comment by mikhailshilkov** [comment]
@jkonecki-techfabric Which `Pulumi.Azure` version are you using?

**Comment by scp-mb** [comment]
It doesn't seem https://github.com/pulumi/pulumi-azure/issues/192 has fixed the issue completely, seeing a very similar issue trying to create a new `AppService` using a managed identity and applying `PrincipalId` to an instance of  `Pulumi.Azure.Authorization.Assignment`

```
new Pulumi.Azure.Authorization.Assignment(<redacted>",
            new Pulumi.Azure.Authorization.AssignmentArgs
            {
                PrincipalId = appService.Identity.Apply(x => x.PrincipalId),
                Scope = ResourceGroup.Id,
                RoleDefinitionName = "Contributor"
            });
```

`error: azure:authorization/assignment:Assignment resource 'redacted' has a problem: Missing required property 'principalId'`

**Comment by jkonecki-techfabric** [comment]
@mikhailshilkov 3.28.0

**Comment by mikhailshilkov** [comment]
@scp-mb Could you add a repro to https://github.com/pulumi/pulumi-azure/issues/192? Maybe we should reopen it.

**Comment by mikhailshilkov** [comment]
I reopened pulumi/pulumi-azure#192 let's track there

**Comment by andrewdmoreno** [comment]
Encountered this issue also with `AccessPolicy` from `Pulumi.AzureNextGen` while attempting to get identity from `WebApp` and `WebAppSlot`, also from `Pulumi.AzureNextGen`. 

The following workaround recommended above **worked**.

`app.Identity.Apply(i => i?.PrincipalId ?? Guid.Empty.ToString())`

**Comment by mikhailshilkov** [comment]
@andrewdmoreno I closed this 10 seconds before your comment :)

Could you add your code snippet here?

**Comment by andrewdmoreno** [comment]
@mikhailshilkov Just want to confirm I'm understanding you correctly: Are you asking me to post the above comment in [pulumi/pulumi-azure#192](https://github.com/pulumi/pulumi-azure/issues/192)? Or were you wanting me to provide more info here re: `Pulumi.AzureNextGen`?

**Comment by andrewdmoreno** [comment]
I'm gonna assume in the meantime that you wanted additional code snippet for `Pulumi.AzureNextGen` here. (Some code omitted for brevity).

````C#
            var app = new WebApp("api", new WebAppArgs
            {
                Name = ResourceName("api"),
                Location = _location,
                ResourceGroupName = _resourceGroup,
                ServerFarmId = plan.Id,
                ClientAffinityEnabled = false,
                Identity = ManagedIdentity(),
                SiteConfig = siteConfigArgs,
                Tags = Tags
            });

            var accessPolicies = new InputList<AccessPolicyEntryArgs>
            {
                DefaultSecretsAccessPolicy(servicePrincipalId),
                DefaultSecretsAccessPolicy(app.Identity.Apply(i => i?.PrincipalId ?? Guid.Empty.ToString())),
                DefaultSecretsAccessPolicy(slot.Identity.Apply(i => i?.PrincipalId ?? Guid.Empty.ToString()))
            };

            var vault = CreateKeyVault(accessPolicies);

        ...

        private Vault CreateKeyVault(InputList<AccessPolicyEntryArgs> accessPolicies)
        {
            return new Vault("vault", new VaultArgs
            {
                ResourceGroupName = _resourceGroup,
                Location = _location,
                Properties = new VaultPropertiesArgs
                {
                    Sku = new SkuArgs
                    {
                        Family = "A",
                        Name = "standard",
                    },
                    TenantId = _azureAdTenantId,
                    AccessPolicies = accessPolicies
                },
                VaultName = _keyVaultName
            });
        }

        private AccessPolicyEntryArgs DefaultSecretsAccessPolicy(Input<string> objectId) => new AccessPolicyEntryArgs
        {
            ObjectId = objectId,
            Permissions = new PermissionsArgs
            {
                Secrets = _secretPermissions,
            },
            TenantId = _azureAdTenantId,
        };
````

I don't have the exact error message this moment, but it was complaining about the access policies with index `[1]` and `[2]` in the `accessPolicies` list.

**Comment by mikhailshilkov** [comment]
I double-checked and there's nothing we can do right now in the provider. The broad issue is tracked in https://github.com/pulumi/pulumi/issues/5864 so I'll close this one - please track there.

**Comment by mikhailshilkov** [comment]
The fix will be released in version 0.3.2

**Comment by mikhailshilkov** [comment]
While this error doesn't happen during the initial preview, it does happen during an update. For instance, if `WebApp` in the example above changes, I get an error `azure-nextgen:keyvault/latest:Vault resource 'vault' has a problem: missing required property 'properties.accessPolicies[0].objectId'`
